### ISSUE TITLE
can't see my fake Xiaomi temp sensor

### ISSUE THREAD

**Issue by vbelloir** [custom-components/ble_monitor#222]
Hi team,

I planned to develop my own BLE sensor, giving temperature and humidity, based on a nRF52 chip.
I make my own sensor, just for fun.

I managed to deal with service datas, and get a advertising very similar to the one given in https://github.com/custom-components/ble_monitor/blob/master/protocol.md
```

> HCI Event: LE Meta Event (0x3e) plen 37                   #31 [hci0] 2.914649
      LE Advertising Report (0x02)
        Num reports: 1
        Event type: Connectable undirected - ADV_IND (0x00)
        Address type: Random (0x01)
        Address: FA:BE:32:DD:0D:36 (Static)
        Data length: 25
        Flags: 0x06
          LE General Discoverable Mode
          BR/EDR Not Supported
        Service Data (UUID 0xfe95): 5020aa0105360ddd32befa0d1004b2007502
        RSSI: -46 dBm (0xd2)
```

The only difference I can notice is the address type. 

I installed ble_monitor through HACS, on a Home installation, in a python virtual env on a ubuntu server.

Here is the HA logs output:
```

2020-12-27 21:47:23 DEBUG (MainThread) [custom_components.ble_monitor.config_flow] async_step_init user_input: None
2020-12-27 21:47:23 DEBUG (MainThread) [custom_components.ble_monitor.config_flow] async_step_init (before): {'hci_interface': [0], 'period': 60, 'discovery': True, 'active_scan': False, 'report_unknown': True, 'batt_entities': False, 'rounding': True, 'decimals': 1, 'log_spikes': False, 'use_median': False, 'restore_state': False, 'devices': [{'mac': 'FA:BE:32:DD:0D:36', 'temperature_unit': '°C', 'name': 'FABE32DD0D36'}], 'is_flow': True}
2020-12-27 21:47:30 DEBUG (MainThread) [custom_components.ble_monitor.config_flow] async_step_init user_input: {'hci_interface': ['0'], 'period': 60, 'discovery': True, 'active_scan': False, 'report_unknown': True, 'batt_entities': False, 'rounding': True, 'decimals': 1, 'log_spikes': False, 'use_median': False, 'restore_state': False, 'devices': 'FA:BE:32:DD:0D:36'}
2020-12-27 21:47:30 DEBUG (MainThread) [custom_components.ble_monitor.config_flow] async_step_init (after): {'hci_interface': ['0'], 'period': 60, 'discovery': True, 'active_scan': False, 'report_unknown': True, 'batt_entities': False, 'rounding': True, 'decimals': 1, 'log_spikes': False, 'use_median': False, 'restore_state': False, 'devices': 'FA:BE:32:DD:0D:36'}
2020-12-27 21:47:32 DEBUG (MainThread) [custom_components.ble_monitor.config_flow] async_step_add_device: {'mac': 'FA:BE:32:DD:0D:36', 'temperature_unit': '°C', 'encryption_key': ''}
2020-12-27 21:47:42 DEBUG (MainThread) [custom_components.ble_monitor.binary_sensor] 0 MiBeacon BLE ADV messages processed for 0 binary sensor device(s) total. Priority queue = 0
2020-12-27 21:47:42 DEBUG (MainThread) [custom_components.ble_monitor.sensor] 0 MiBeacon BLE ADV messages processed for 1 measuring device(s).
2020-12-27 21:47:42 DEBUG (Thread-8) [custom_components.ble_monitor] HCIdump thread: main event_loop stopped, finishing
2020-12-27 21:47:42 DEBUG (Thread-8) [custom_components.ble_monitor] HCIdump thread: Scanning will be restarted
2020-12-27 21:47:42 DEBUG (Thread-8) [custom_components.ble_monitor] 2618 HCI events processed for previous period.
2020-12-27 21:47:42 DEBUG (Thread-8) [custom_components.ble_monitor] HCIdump thread: Run
2020-12-27 21:47:42 DEBUG (Thread-8) [custom_components.ble_monitor] HCIdump thread: connected to hci0
2020-12-27 21:47:42 DEBUG (Thread-8) [custom_components.ble_monitor] HCIdump thread: start main event_loop
2020-12-27 21:48:42 DEBUG (MainThread) [custom_components.ble_monitor.binary_sensor] 0 MiBeacon BLE ADV messages processed for 0 binary sensor device(s) total. Priority queue = 0
2020-12-27 21:48:42 DEBUG (MainThread) [custom_components.ble_monitor.sensor] 0 MiBeacon BLE ADV messages processed for 1 measuring device(s).
2020-12-27 21:48:42 DEBUG (Thread-8) [custom_components.ble_monitor] HCIdump thread: main event_loop stopped, finishing
2020-12-27 21:48:42 DEBUG (Thread-8) [custom_components.ble_monitor] HCIdump thread: Scanning will be restarted
2020-12-27 21:48:42 DEBUG (Thread-8) [custom_components.ble_monitor] 2473 HCI events processed for previous period.
2020-12-27 21:48:42 DEBUG (Thread-8) [custom_components.ble_monitor] HCIdump thread: Run
2020-12-27 21:48:42 DEBUG (Thread-8) [custom_components.ble_monitor] HCIdump thread: connected to hci0
2020-12-27 21:48:42 DEBUG (Thread-8) [custom_components.ble_monitor] HCIdump thread: start main event_loop
2020-12-27 21:49:42 DEBUG (MainThread) [custom_components.ble_monitor.binary_sensor] 0 MiBeacon BLE ADV messages processed for 0 binary sensor device(s) total. Priority queue = 0
2020-12-27 21:49:42 DEBUG (MainThread) [custom_components.ble_monitor.sensor] 0 MiBeacon BLE ADV messages processed for 1 measuring device(s).
2020-12-27 21:49:42 DEBUG (Thread-8) [custom_components.ble_monitor] HCIdump thread: main event_loop stopped, finishing
2020-12-27 21:49:42 DEBUG (Thread-8) [custom_components.ble_monitor] HCIdump thread: Scanning will be restarted
2020-12-27 21:49:42 DEBUG (Thread-8) [custom_components.ble_monitor] 2281 HCI events processed for previous period.
2020-12-27 21:49:42 DEBUG (Thread-8) [custom_components.ble_monitor] HCIdump thread: Run
2020-12-27 21:49:42 DEBUG (Thread-8) [custom_components.ble_monitor] HCIdump thread: connected to hci0
2020-12-27 21:49:42 DEBUG (Thread-8) [custom_components.ble_monitor] HCIdump thread: start main event_loop

```

The sensor is present (FA:BE:32:DD:0D:36), but not data appears in HA.
![image](https://user-images.githubusercontent.com/13942805/103179697-dc174300-488e-11eb-8a1d-6cb3dc7fad08.png)

I probably missed something, but where?

Thanks for your job and support.
Vincent

**Comment by vbelloir** [comment]
the report_unknown option is selected, but I can find home-assistant.log
[home-assistant.log](https://github.com/custom-components/ble_monitor/files/5745203/home-assistant.log)

**Comment by Ernst79** [comment]
You will have to comment out the sensor you try to mimic in const.py to make it show up as unknown sensor. Perhaps that will give more clues

**Comment by vbelloir** [comment]
I didn't edit the const.py file.
My device is declared as a LYWSDCGQ, like in the (protocol.md)[https://github.com/custom-components/ble_monitor/blob/master/protocol.md], with 01AA, that's why it is not unkown... sorry for that.

I change it for 01AB, wich is not used in const.py.

btmon output is now :
```
> HCI Event: LE Meta Event (0x3e) plen 37                 #620 [hci0] 56.064012
      LE Advertising Report (0x02)
        Num reports: 1
        Event type: Connectable undirected - ADV_IND (0x00)
        Address type: Random (0x01)
        Address: FA:BE:32:DD:0D:36 (Static)
        Data length: 25
        Flags: 0x06
          LE General Discoverable Mode
          BR/EDR Not Supported
        Service Data (UUID 0xfe95): 5020ab0105360ddd32befa0d1004b2007502
        RSSI: -45 dBm (0xd3)
```

but there is still not report on home-assistant.log

**Comment by Ernst79** [comment]
That is also possible. 
As far as I can see , the service data looks ok. Can you try to delete the device in your configuration?

**Comment by Magalex2x14** [comment]
@vbelloir the type of address is not important yet. I say "yet", because in the future, after the implementation of the device tracker, advertisements with random addresses will be dropped.
The parser checks the integrity of the advertisement on many grounds (there are cases when strange, incorrect packages slip through). Among them:
- 1695fe bytes must be present in a specific location
- the address in the body of the payload must match the address from which the advertisement actually came
- the actual length of the message must match the one specified in the advertisement
- the value of the message counter must not be equal to the value from the previous advertisement
- and something else (I don't remember all the checks now, I have to look at the code)

If you do not receive messages about an unknown device in the log, they are filtered out by some of the signs even before checking the device type (before the `sensor_type, binary_data = XIAOMI_TYPE_DICT[data[xiaomi_index + 5:xiaomi_index + 7]]` of the `parse_raw_message` method).

**Comment by vbelloir** [comment]
@Ernst79 , done, without any success.

@Magalex2x14 : one of the issue could be  `frame cnt` which is always 0, because I always send the same advertising packet.
I will write and updater tomorrow, that will increase this field.

When yo write, 1695fe, you mean 0x95FE, which correspond to the Xiami UUID isn't it?

**Comment by Magalex2x14** [comment]
The parser checks exactly the presence of 0x1695FE, which means
- advertisement structure type: Service data (0x16)
- Xiaomi 16bit UUID 0x95FE
later, when I have more free time, I will supplement `protocol.md` with my knowledge

**Comment by vbelloir** [comment]
Hi,
as you can see in the btmon output, we can expect to get th 0x1695FE, because we got a `Service Data (UUID 0xfe95)`, as in protocol.md.

with btmon I can't get complet raw data, but on Android with nRFConnect, I can:
![image](https://user-images.githubusercontent.com/13942805/103200801-9b4f1680-48ee-11eb-8608-3260d544e199.png)

And even if we can see that the service seems ok, the adv packet format is not the same as in `protocol.md`. I don't real Xaomi devices, so I can't compare.

Looking at https://www.bluetooth.com/specifications/assigned-numbers/generic-access-profile/
Type 01 is flags and flag is 0x06, like in btmon output.
Type 16 is «Service Data», and start with UUID 95FE
Type 17 is «Public Target Address»

Another to see it is here
![image](https://user-images.githubusercontent.com/13942805/103201220-ac4c5780-48ef-11eb-94b7-8ea9caab8253.png)

Two kind of issue possible:
- frame count is not changing, and this is the issue with the parser,
- the whaole advertisement format is not the same, and the issue is bigger.

**Comment by Ernst79** [comment]
I've installed nRf connect for you. The info from the same device as in `protocol.md`. The raw data seems to be quite a bit longer, I also see a second service data part. 

![Screenshot_20201228-101316_nRF Connect](https://user-images.githubusercontent.com/24375847/103204881-97c08d00-48f8-11eb-8d9b-7ba2e84629fd.jpg)

![Screenshot_20201228-101421_nRF Connect](https://user-images.githubusercontent.com/24375847/103204912-a8710300-48f8-11eb-8db4-db2d2adcba8f.jpg)

**Comment by vbelloir** [comment]
Thanks @Ernst79 !
as far as I can see, service data for 0x95FE service seems to be ok.  

I am not very familiar with python, but I tried to instrument ble_monitor parser to see what part of the advertising packet is not compliant with it.
I found one mistake from my side. Great.
Now, I can get :  
`
2020-12-29 10:48:08 INFO (Thread-3) [custom_components.ble_monitor] BLE ADV from UNKNOWN: RSSI: -81, MAC: FABE32DD0D36, ADV: 043e2502010001360ddd32befa19020106151695fe5020ab0105360ddd32befa0d1004b2007502af`
 When I used a fake new device type of 0X01AB.
My advertising packet is now compliant with ble_monitor_parser. Great too!!


![image](https://user-images.githubusercontent.com/13942805/103275491-952b6980-49c4-11eb-8455-e10594c5a32c.png)

And now, I got my fake temp and humidity value in HA!

Thanks for your support guys!
